src/lib/Syntax.hs

```diff
@@ -552,9 +563,16 @@ data BaseType = Scalar  ScalarBaseType
 
 data Device = CPU | GPU  deriving (Show, Eq, Ord, Generic)
 data AddressSpace = Stack | Heap Device     deriving (Show, Eq, Ord, Generic)
-data PtrOrigin = DerivedPtr | AllocatedPtr  deriving (Show, Eq, Ord, Generic)
+data PtrOrigin = DerivedPtr | AllocatedPtr  deriving (Show, Ord, Generic)
 type PtrType = (PtrOrigin, AddressSpace, BaseType)
 
+instance Eq PtrOrigin where
+  -- XXX: this is a hack. We expose pointer operations to the surface language
+  -- but we don't yet expose the derived/allocated distinction, and they get
+  -- mixed up when we use ops like ptrOffset.
+  _ == _ = True
```

Comment:
I don't get it. Why would this `Eq` instance have any effect on the semantics of the surface language? We definitely shouldn't merge it as is

Reply:
As discussed offline, let's just get rid of the derived/allocated distinction. We never ended up using it.